File: new/valid_anagram_frequency_map.cpp
class Solution {
public:
    bool isAnagram(string s, string t) {
        // if both strings don't have same length, they cannot be anagram
        // create frequency map for first string
        // go throgugh the frequency map for every char in second string
        // if by the end the frequency map has any other number other than 0
        // they cannot be anagram return false
        // if all the characters have 0, return true
        if (s.size() != t.size()) {
            return false;
        }
        map<char, int> frequency_counter;
        for (int i = 0; i < s.size(); i++) {
            if (frequency_counter.count(s[i]) > 0) {
                frequency_counter[s[i]] += 1;
            } else {
                frequency_counter[s[i]] = 1;
            }
        }

        for (int j = 0; j < t.size(); j++) {
            if (frequency_counter.count(t[j]) > 0) {
                frequency_counter[t[j]] -= 1;
            } else {
                return false;
            }
        }

        for (auto& pair: frequency_counter) {
            if (pair.second != 0) {
                return false;
            }
        }

        return true;
        // time complexity: O(n)
        // space complexity: O(1) if fixed alphabets if any type of characters O(n)
    }
};

File: new/python/valid_anagram_frequency_map.py
class Solution:
    def isAnagram(self, s: str, t: str) -> bool:
        # check if they have same length if not they cannot be anagrams
        # if same length, create frequency map for first string
        # map the frequency of first string, go through it
        # go through the second string and reduce the count of frequency of frequency map
        # if the map after the end of the process has frequency map of all 0s it is anagram
        # if not they are not anagram

        if len(s) != len(t):
            return False

        frequency_counter = {}
        for i in range(0, len(s)):
            if s[i] in frequency_counter:
                frequency_counter[s[i]] += 1
            else:
                frequency_counter[s[i]] = 1

        # we don't even need this for loop we can do the step above itself.
        for j in range(0, len(s)):
            if t[j] in frequency_counter:
                frequency_counter[t[j]] -= 1
            else:
                return False

        for k in frequency_counter:
            if frequency_counter[k] != 0:
                return False

        return True
        # time complexity: O(n)
        # space complexity: O(1) if fixed alphabets if any type of characters O(n)
